src/connector/src/sink/kinesis.rs

```diff
@@ -191,39 +189,176 @@ impl KinesisSinkWriter {
+
             async move {
-                let builder = self.builder.expect("should not be None");
-                let context_fmt = format!(
-                    "failed to put record to {}",
-                    builder
-                        .get_stream_name()
-                        .as_ref()
-                        .expect("should have set stream name")
-                );
-                Retry::spawn(
-                    ExponentialBackoff::from_millis(100).map(jitter).take(3),
-                    || builder.clone().send(),
-                )
-                .await
-                .with_context(|| context_fmt.clone())
-                .map_err(SinkError::Kinesis)?;
+                // From the doc of `put_records`:
+                // Each PutRecords request can support up to 500 records. Each record in the request can be as large as 1 MiB,
+                // up to a limit of 5 MiB for the entire request, including partition keys. Each shard can support writes up to
+                // 1,000 records per second, up to a maximum data write total of 1 MiB per second.
+
+                const MAX_RECORD_COUNT: usize = 500;
+                const MAX_SINGLE_RECORD_PAYLOAD_SIZE: usize = 1 << 20;
+                const MAX_TOTAL_RECORD_PAYLOAD_SIZE: usize = 5 * (1 << 20);
+                // Allow at most 3 times of retry when not making any progress to avoid endless retry
+                const MAX_NO_PROGRESS_RETRY_COUNT: usize = 3;
+
+                let mut remaining_no_progress_retry_count = MAX_NO_PROGRESS_RETRY_COUNT;
+                let total_count = self.entries.len();
+                let mut start_idx = 0;
+
+                let mut throttle_delay = None;
+
+                while start_idx < total_count {
+                    // 1. Prepare the records to be sent
+
+                    // The maximum possible number of records that can be sent in this iteration.
+                    // Can be smaller than this number of the total payload size exceeds `MAX_TOTAL_RECORD_PAYLOAD_SIZE`
+                    let max_record_count = min(MAX_RECORD_COUNT, total_count - start_idx);
+                    let mut records = Vec::with_capacity(max_record_count);
+                    let mut total_payload_size = 0;
+                    for i in start_idx..(start_idx + max_record_count) {
+                        let (record, size) = &self.entries[i];
+                        if *size >= MAX_SINGLE_RECORD_PAYLOAD_SIZE {
+                            warn!(
+                                size,
+                                partition = record.partition_key,
+                                "encounter a large single record"
+                            );
+                        }
+                        if total_payload_size + *size < MAX_TOTAL_RECORD_PAYLOAD_SIZE {
+                            total_payload_size += *size;
+                            records.push(record.clone());
```

Comment:
The need of `clone` here feels like a API design flaw in the AWS auto-generated SDK. I don't think we can do anything but just point it out.

Impact:
Every single message written out is cloned at least once.

Root cause:
AWS SDK takes the ownership away but does not give it back for failure retry.